File: backend/app/schemas/billing.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class EntitlementOut(BaseModel):
    id: UUID
    entitlement_key: str
    source: str
    expires_at: datetime | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ManualGrantRequest(BaseModel):
    user_id: UUID
    entitlement_key: str
    source: str = "manual_grant"


class PremiumStatus(BaseModel):
    is_premium: bool
    entitlement: str | None = None

File: frontend/src/api/billing.ts
import client from './client';

export interface PremiumStatus {
  is_premium: boolean;
  entitlement: string | null;
}

export interface Entitlement {
  id: string;
  entitlement_key: string;
  source: string;
  expires_at: string | null;
  created_at: string;
}

export async function getPremiumStatus(): Promise<PremiumStatus> {
  const res = await client.get('/billing/status');
  return res.data;
}

export async function getMyEntitlements(): Promise<Entitlement[]> {
  const res = await client.get('/billing/entitlements');
  return res.data;
}

export async function adminGrantEntitlement(
  userId: string,
  entitlementKey = 'ads_removed',
  source = 'manual_grant',
): Promise<Entitlement> {
  const res = await client.post('/billing/grant', {
    user_id: userId,
    entitlement_key: entitlementKey,
    source,
  });
  return res.data;
}

export async function adminRevokeEntitlement(
  userId: string,
  entitlementKey = 'ads_removed',
): Promise<void> {
  await client.delete(`/billing/grant/${userId}/${entitlementKey}`);
}
